File: 22.reactor-reboot/scala/part2.scala

import scala.io.StdIn
import scala.collection.mutable.ListBuffer
import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet
import scala.math.BigInt

type Box = ((Int, Int), (Int, Int), (Int, Int))

object part2 {
    def main(args: Array[String]) = {
        var seq = ListBuffer[(Boolean, Box)]()
        var line = ""
        while ({line = StdIn.readLine(); line != null}) {
            var split = line.split(" ")
            var on = split(0) == "on"
            var arr = split(1).split(",").map(s => s.split("=")(1).split("\\.\\.").map(v => v.toInt))
            var box = ((arr(0)(0), arr(0)(1)), (arr(1)(0), arr(1)(1)), (arr(2)(0), arr(2)(1)))
            seq += ((on, box))
        }
        var reactor = HashMap[Box, BigInt]().withDefaultValue(0)
        for ((on, bx) <- seq) {
            var updates = HashMap[Box, BigInt]().withDefaultValue(0)
            for ((xb, sgn) <- reactor) {
                var isx0 = bx(0)(0).max(xb(0)(0)); var isx1 = bx(0)(1).min(xb(0)(1))
                var isy0 = bx(1)(0).max(xb(1)(0)); var isy1 = bx(1)(1).min(xb(1)(1))
                var isz0 = bx(2)(0).max(xb(2)(0)); var isz1 = bx(2)(1).min(xb(2)(1))
                var box = ((isx0, isx1), (isy0, isy1), (isz0, isz1))
                if (isx0 <= isx1 && isy0 <= isy1 && isz0 <= isz1) {
                    updates(box) -= sgn
                }
            }
            for ((xb, sgn) <- updates) {
                reactor(xb) += sgn
            }
            if (on) {
                reactor(bx) += 1
            }
        }
        var sum = BigInt(0)
        for ((xb, sgn) <- reactor) {
            sum += sgn * (xb(0)(1) - xb(0)(0) + 1) * (xb(1)(1) - xb(1)(0) + 1) * (xb(2)(1) - xb(2)(0) + 1)
        }
        println("Result: " + sum)
    }
}

File: 22.reactor-reboot/py/part2.py

import sys

seq = []
for line in sys.stdin.read().strip().split('\n'):
    on, pos = line.split(' ')
    iv = tuple(tuple(int(v) for v in l.split('=')[1].split('..')) for l in line.split(','))
    seq.append((on == 'on', iv))

def splitIntersection(i0, i1):
    xs, ys, zs = [sorted(i0[i] + i1[i]) for i in range(3)]
    splits = [
        ((xs[1], xs[2]), (ys[1], ys[2]), (zs[2] + 1, i1[2][1])),
        ((xs[1], xs[2]), (ys[1], ys[2]), (i1[2][0], zs[1] - 1)),
        ((xs[1], xs[2]), (i1[1][0], ys[1] - 1), i1[2]),
        ((xs[1], xs[2]), (ys[2] + 1, i1[1][1]), i1[2]),
        ((i1[0][0], xs[1] - 1), i1[1], i1[2]),
        ((xs[2] + 1, i1[0][1]), i1[1], i1[2]),
    ]
    return filter(lambda v: v[0][1] >= v[0][0] and v[1][1] >= v[1][0] and v[2][1] >= v[2][0], splits)

reactor = set()

for on, iv in seq:
    overlaps = []
    for i in reactor:
        if (
            iv[0][0] <= i[0][1] and iv[0][1] >= i[0][0]
            and iv[1][0] <= i[1][1] and iv[1][1] >= i[1][0]
            and iv[2][0] <= i[2][1] and iv[2][1] >= i[2][0]
        ):
            overlaps.append(i)
    reactor.difference_update(overlaps)
    for ov in overlaps:
        reactor.update(splitIntersection(iv, ov))
    if on:
        reactor.add(iv)

count = sum(map(lambda i: (i[0][1] - i[0][0] + 1) * (i[1][1] - i[1][0] + 1) * (i[2][1] - i[2][0] + 1), reactor))

print("Result:",  count)

